### PR TITLE
agent: add /v1/coordianate/node/:node endpoint

### PR DESCRIPTION
**Pull request by magiconair**
This patch adds a /v1/coordinate/node/:node endpoint to get the network
coordinates for a single node in the network.

Since Consul Enterprise supports network segments it is still possible
to receive mutiple entries for a single node - one per segment.

### CODE REVIEW

File: agent/consul/coordinate_endpoint.go [hunk 1]
```diff
@@ -200,3 +200,45 @@ func (c *Coordinate) ListNodes(args *structs.DCSpecificRequest, reply *structs.I
 			return nil
 		})
 }
+
+// ListNodes returns the list of nodes with their raw network coordinates (if no
```
Comment (magiconair) [inline agent/consul/coordinate_endpoint.go]: Comment does not match.

File: agent/consul/coordinate_endpoint_test.go [hunk 2]
```diff
@@ -502,3 +502,147 @@ node "foo" {
...
+	defer os.RemoveAll(dir1)
+	defer s1.Shutdown()
+
+	codec := rpcClient(t, s1)
+	defer codec.Close()
+	testrpc.WaitForLeader(t, s1.RPC, "dc1")
+
+	// Register some nodes.
+	nodes := []string{"foo", "bar"}
+	for _, node := range nodes {
```
Comment (magiconair) [inline agent/consul/coordinate_endpoint_test.go]: Maybe create a `registerNodes` helper in a separate commit first?

File: api/coordinate.go [hunk 3]
```diff
@@ -66,3 +66,24 @@ func (c *Coordinate) Nodes(q *QueryOptions) ([]*CoordinateEntry, *QueryMeta, err
 	}
 	return out, qm, nil
 }
+
+// Node is used to return the coordinates of a single in the LAN pool.
+func (c *Coordinate) Node(node string, q *QueryOptions) ([]*CoordinateEntry, *QueryMeta, error) {
```
Comment (magiconair) [inline api/coordinate.go]: Right, I've forgotten about the API client.

File: agent/consul/state/coordinate.go [hunk 4]
```diff
@@ -42,21 +42,25 @@ func (s *Restore) Coordinates(idx uint64, updates structs.Coordinates) error {
 
 // Coordinate returns a map of coordinates for the given node, indexed by
 // network segment.
-func (s *Store) Coordinate(node string) (lib.CoordinateSet, error) {
+func (s *Store) Coordinate(node string, ws memdb.WatchSet) (uint64, lib.CoordinateSet, error) {
 	tx := s.db.Txn(false)
 	defer tx.Abort()
 
+	// Get the table index.
+	idx := maxIndexTxn(tx, "coordinates")
```
Comment (magiconair) [inline agent/consul/state/coordinate.go]: `s/idx/tableIdx/` and drop the comment

File: agent/consul/rtt.go [hunk 5]
```diff
@@ -22,7 +22,7 @@ func (s *Server) newNodeSorter(cs lib.CoordinateSet, nodes structs.Nodes) (sort.
 	state := s.fsm.State()
 	vec := make([]float64, len(nodes))
 	for i, node := range nodes {
-		other, err := state.Coordinate(node.Node)
+		_, other, err := state.Coordinate(node.Node, nil)
```
Comment (magiconair) [inline agent/consul/rtt.go]: Why doesn't this panic since `Coordinate(` doesn't handle `ws == nil`?

Comment (kyhavlov) [reply agent/consul/rtt.go]: memdb.WatchSet itself has the nil checks

File: agent/consul/coordinate_endpoint.go [hunk 6]
```diff
@@ -200,3 +200,45 @@ func (c *Coordinate) ListNodes(args *structs.DCSpecificRequest, reply *structs.I
...
+	// Fetch the ACL token, if any, and enforce the node policy if enabled.
+	rule, err := c.srv.resolveToken(args.Token)
+	if err != nil {
+		return err
+	}
+	if rule != nil && c.srv.config.ACLEnforceVersion8 {
+		// We don't enforce the sentinel policy here, since at this time
+		// sentinel only applies to creating or updating node or service
+		// info, not updating coordinates.
+		if !rule.NodeWrite(args.Node, nil) {
```
[the remark this answers is not in the document]
Comment (slackpad) [reply agent/consul/coordinate_endpoint.go]: (can you plz. also delete the comment this was copied from)?

File: agent/consul/coordinate_endpoint_test.go [hunk 7]
```diff
@@ -502,3 +479,147 @@ node "foo" {
...
+		}
+		if len(resp.Coordinates) != 1 ||
+			resp.Coordinates[0].Node != "foo" {
+			r.Fatalf("bad: %v", resp.Coordinates)
+		}
+		verify.Values(t, "", resp.Coordinates[0].Coord, arg1.Coord) // foo
+	})
+}
+
+func TestCoordinate_Node_ACLDeny(t *testing.T) {
```
Comment (slackpad) [inline agent/consul/coordinate_endpoint_test.go]: This test never ended up calling `Coordinate.Node`

File: agent/consul/state/coordinate.go [hunk 8]
```diff
@@ -42,21 +42,24 @@ func (s *Restore) Coordinates(idx uint64, updates structs.Coordinates) error {
 
 // Coordinate returns a map of coordinates for the given node, indexed by
 // network segment.
-func (s *Store) Coordinate(node string) (lib.CoordinateSet, error) {
+func (s *Store) Coordinate(node string, ws memdb.WatchSet) (uint64, lib.CoordinateSet, error) {
```
Comment (slackpad) [inline agent/consul/state/coordinate.go]: Would be good to add a little coverage for `ws` and the returned index in the unit tests.

File: agent/coordinate_endpoint.go [hunk 9]
```diff
@@ -85,22 +86,53 @@ func (s *HTTPServer) CoordinateNodes(resp http.ResponseWriter, req *http.Request
...
+	}
+
+	var out structs.IndexedCoordinates
+	defer setMeta(resp, &out.QueryMeta)
+	if err := s.agent.RPC("Coordinate.Node", &args, &out); err != nil {
+		sort.Sort(&sorter{out.Coordinates})
+		return nil, err
+	}
+
+	return filterCoordinates(req, node, out.Coordinates), nil
```
Comment (slackpad) [inline agent/coordinate_endpoint.go]: What do you think about a 404 if there's no coordinate available?

File: agent/coordinate_endpoint.go [hunk 10]
```diff
@@ -85,22 +86,53 @@ func (s *HTTPServer) CoordinateNodes(resp http.ResponseWriter, req *http.Request
...
+}
+
+func filterCoordinates(req *http.Request, node string, in structs.Coordinates) structs.Coordinates {
```
Comment (slackpad) [inline agent/coordinate_endpoint.go]: Do we need to filter by node name? This seems like maybe a leftover from when this was built upon the existing endpoint?

File: website/source/api/coordinate.html.md [hunk 11]
```diff
@@ -122,3 +122,55 @@ $ curl \
 In **Consul Enterprise**, this may include multiple coordinates for the same node,
 each marked with a different `Segment`. Coordinates are only compatible within the same
 segment.
+
+## Read LAN Coordinates for a node
+
+This endpoint returns the LAN network coordinates for all nodes in a given
```
Comment (slackpad) [inline website/source/api/coordinate.html.md]: This is stale.

File: website/source/api/coordinate.html.md [hunk 12]
```diff
@@ -122,3 +122,55 @@ $ curl \
...
+
+| Blocking Queries | Consistency Modes | ACL Required |
+| ---------------- | ----------------- | ------------ |
+| `YES`            | `all`             | `node:read`  |
+
+### Parameters
+
+- `dc` `(string: "")` - Specifies the datacenter to query. This will default to
+  the datacenter of the agent being queried. This is specified as part of the
+  URL as a query parameter.
```
Comment (slackpad) [inline website/source/api/coordinate.html.md]: It also supports `?segment`

File: agent/coordinate_endpoint.go [hunk 13]
```diff
@@ -85,22 +86,56 @@ func (s *HTTPServer) CoordinateNodes(resp http.ResponseWriter, req *http.Request
...
+	node := strings.TrimPrefix(req.URL.Path, "/v1/coordinate/node/")
+	args := structs.NodeSpecificRequest{Node: node}
+	if done := s.parse(resp, req, &args.Datacenter, &args.QueryOptions); done {
+		return nil, nil
+	}
+
+	var out structs.IndexedCoordinates
+	defer setMeta(resp, &out.QueryMeta)
+	if err := s.agent.RPC("Coordinate.Node", &args, &out); err != nil {
+		sort.Sort(&sorter{out.Coordinates})
```
Comment (slackpad) [inline agent/coordinate_endpoint.go]: Drop the sort in here.

File: agent/coordinate_endpoint_test.go [hunk 14]
```diff
@@ -140,3 +140,106 @@ func TestCoordinate_Nodes(t *testing.T) {
...
+		Datacenter: "dc1",
+		Node:       "bar",
+		Coord:      coordinate.NewCoordinate(coordinate.DefaultConfig()),
+	}
+	if err := a.RPC("Coordinate.Update", &arg2, &out); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	time.Sleep(300 * time.Millisecond)
+
+	// Query back and check the nodes are present and sorted correctly.
```
Comment (slackpad) [inline agent/coordinate_endpoint_test.go]: "sorted correctly" is a stale comment - you can drop that part

File: api/coordinate_test.go [hunk 15]
```diff
@@ -3,6 +3,8 @@ package api
 import (
 	"testing"
```
Comment (slackpad) [inline api/coordinate_test.go]: Extra line here.

**Pull request merged**